### ISSUE TITLE
Best practices when using `rsample` in modelling packages

### ISSUE THREAD

**Issue by alexpghayes** [topepo/rsample#33]
I'd like to convince a professor of mine to use `rsample` in a modelling package she's working on. The hope would be to deal with one new abstraction from rsample and gain several new types of resampling in the process. 

The model currently has two options for users to specify data (with a recipes interface coming soon!):

- a data matrix `X` together with a response vector `y`
- a formula `formula` together with a data frame `data`

The user specifies how many folds to use and potentially how many times to repeat the cross validation via some arguments to the model fitting function.

My initial thought was to extend the formula interface to allow the `data` argument to be an `rset`. But I don't think there's a clean parallel for the matrix interface since `rset` generators need a data frame.

An alternative might be for to write a resampling *specification* in the vein of recipes that could then be applied to either a matrix or a data frame. This is more appealing to me since it keeps the data entry and model calibration options separate.  In our case at least, this model calibration specification could come as a package deal including about resampling, loss function, loss strategy (i.e. min RMSE vs sparsest within 1-se) and some other custom tuning knobs.

The workflow might then look something like this

``` r
trained_model <- 
  model(formula/X,
        data/y,
        calibration = new_calibration(method = "vfold_cv",
                                      v = 5,
                                      loss = "rmse",
                                      rule = "1-se",
                                      custom_option = "blah"),
        ...)
```

What do you think is the best direction to go here?

(Aside: is there a way to extract `train`/`test` indices from an `rsplit` object? This would let us reuse a lot of existing code).

**Comment by topepo** [comment]
> But I don't think there's a clean parallel for the matrix interface since rset generators need a data frame.

That's correct. We tend to see data frames as the default format that can be converted to a matrix as needed. You could always temporarily convert the matrix to a data frame for the purposes of resampling. 

Model formulas aren't perfect but they are very expressive. Also, as someone whose job was modeling, it is really frustrating when a function takes a matrix (`x`/`/y`) interface since you have to do all of the data manipulations manually (e.g. dummy variables, transformations, subsets etc). `recipes` is suppose to solve a lot of those issues. 

You could parameterize the calibration function as

```r
new_calibration(
  method = "vfold_cv",
  ..., # to pass to the `rsample` function, don't hardcode options
  control = list(loss , rule, custom_option) # makes it a little more concise. 
)
```

> (Aside: is there a way to extract train/test indices from an rsplit object? This would let us reuse a lot of existing code).

Yes, see [`analysis` and `assessment`](https://topepo.github.io/rsample/reference/as.data.frame.rsplit.html) (which I now realize has no examples): 

```r
> folds <- vfold_cv(mtcars)
> folds$splits[[1]] %>% analysis() %>% str()
'data.frame':	28 obs. of  11 variables:
 $ mpg : num  21 21 22.8 21.4 18.1 14.3 24.4 22.8 19.2 16.4 ...
 $ cyl : num  6 6 4 6 6 8 4 4 6 8 ...
 $ disp: num  160 160 108 258 225 ...
 $ hp  : num  110 110 93 110 105 245 62 95 123 180 ...
 $ drat: num  3.9 3.9 3.85 3.08 2.76 3.21 3.69 3.92 3.92 3.07 ...
 $ wt  : num  2.62 2.88 2.32 3.21 3.46 ...
 $ qsec: num  16.5 17 18.6 19.4 20.2 ...
 $ vs  : num  0 0 1 1 1 0 1 1 1 0 ...
 $ am  : num  1 1 1 0 0 0 0 0 0 0 ...
 $ gear: num  4 4 4 3 3 3 4 4 4 3 ...
 $ carb: num  4 4 1 1 1 4 2 2 4 3 ...
```
